Develop self play self play controller trainer

Danielhp95 opened the pull request:
Introduced `SelfPlayTrainerController` as a child class of TrainerController. It's purpose is to keep a `tf.train.Saver` per `ghost` brain so that it can save checkpoints of these brains to serve as policies for future sampling. This is the core of our self-play mechanism

Pete-York committed:
Create skeleton of self_play_trainer_controller.py

Pete-York committed:
Begin to add structure to self_play_trainer_controller.py

Danielhp95 committed:
Introduced savers for all ghost brains in SelfPlayTrainerController.
TrainerController further refactored, save_models_and_checkpoints
function introduced what can potentially be saved on every step

Pete-York commented on python/unitytrainers/self_play_trainer_controller.py:
```diff
@@ -0,0 +1,64 @@
+
+        # TODO add opci, delta and opponent save frequency
+
+    def _initialize_trainers(self, trainer_config, sess):
```
I'm not sure where this and `initialize_ghost_trainer_savers(self, ghost_trainers)` are called. Should they be called in the constructor?

gaoyuankidult replied:
`_initialize_trainers(self, ghost_trainers)` is called by `initialise_training_environment`, which is in SelfPlayTrainerController. However, I did not see where is `initialise_training_environment` called.  @Danielhp95 Do you put them in the constructor? Maybe it is in the TODO list?

Danielhp95 replied:
the function `_initializer_trainers` is an overriden function, the original one coming from `trainer_controller`. It is called in the `initialize_training_environment`, which in turn is a function that we refactored (previously part of `start_learning`). It should not be part of the constructor, as it takes a session as an argument, and the session isn't created until inside `start_learning`.

Pete-York replied:
ok sounds good to me. The only thing then is, why does it start with `_`

Danielhp95 replied:
Because Python is too cool to care about encapsulation, a method is considered "private" if it starts with an underscore. I personally dislike this convention. It starts with underscore because the original method in the `TrainerController` starts with an underscore.

Pete-York commented on python/unitytrainers/self_play_trainer_controller.py:
```diff
@@ -0,0 +1,64 @@
+            self.logger.info("Saving new opponents behaviour")
+            for brain_name, saver in self.ghost_trainer_savers.items():
+                last_checkpoint = self.ghost_trainers_path + brain_name + '-' + str(global_step) + '.cptk'
+                saver.save(sess, last_checkpoint)
```
Do we mean to save the ghost checkpoints? Somewhat defeats the point of self-play, right? We should be resampling only from the main guy

Danielhp95 replied:
Huh... yeah, you are right. This should really not be here. My bad.

Pete-York reviewed (commented):
There could be some other things I've missed but I hada couple of thoughts

gaoyuankidult commented on python/unitytrainers/self_play_trainer_controller.py:
```diff
@@ -0,0 +1,64 @@
+        if not os.path.exists(self.ghost_trainers_path):
+            os.makedirs(self.ghost_trainers_path)
+
+        # TODO add opci, delta and opponent save frequency
```
I would suggest raising an exception if TODO is not implemented. If we are at the testing phase, then ignore this comment.

Danielhp95 replied:
I will include an exception. The testing phase should come right after this merge is finished, so it makes sense to have an exception here

gaoyuankidult commented on python/unitytrainers/self_play_trainer_controller.py:
```diff
@@ -0,0 +1,64 @@
+    def save_models_and_checkpoints(self, sess, global_step, global_saver):
+        super(SelfPlayTrainerController, self).save_models_and_checkpoints(sess, global_step, global_saver)
+        if global_step % self.save_freq == 0 and global_step != 0 and self.train_model:
+            self.logger.info("Saving new opponents behaviour")
```
Should here be `self.logger.info("Saving new opponent's behaviour")` ?

Danielhp95 replied:
This function will change quite a bit, so the logger will info will change.

gaoyuankidult commented on python/unitytrainers/trainer.py:
```diff
@@ -177,3 +170,17 @@ def write_tensorboard_text(self, key, input_dict):
         except:
             logger.info("Cannot write text summary for Tensorboard. Tensorflow version must be r1.2 or above.")
             pass
```
Should we modify this function in order to save the extra flags or is it also automatically handled by this function?

Danielhp95 committed:
Merge remote-tracking branch 'origin/develop-self-play' into develop-self-play-SelfPlayControllerTrainer

Danielhp95 committed:
Changed save models function to save only the checkpoint with the
variables corresponding to the main brain. Change the initialize
function for self play brains to reflect this.

Danielhp95 committed:
Introduced is_main and is_ghost fields in trainer

Danielhp95 committed:
changed filetering to only filter on trainer.is_main
Renamed a few variables
